src/language-mode.coffee

```diff
@@ -255,12 +255,32 @@ class LanguageMode
       precedingRow = bufferRow - 1
       return currentIndentLevel if precedingRow < 0
 
-    precedingLine = @buffer.lineForRow(precedingRow)
     desiredIndentLevel = @editor.indentationForBufferRow(precedingRow)
-    desiredIndentLevel += 1 if increaseIndentRegex.testSync(precedingLine) and not @editor.isBufferRowCommented(precedingRow)
+    unless @editor.isBufferRowCommented(precedingRow)
+      precedingLine = @buffer.lineForRow(precedingRow)
+      desiredIndentLevel += 1 if increaseNextIndentRegex.testSync(precedingLine)
+
+    decreaseNextIndentRegex = @decreaseNextIndentRegexForScopeDescriptor(scopeDescriptor)
+    decreaseIndentRegex = @decreaseIndentRegexForScopeDescriptor(scopeDescriptor)
+    return desiredIndentLevel unless decreaseNextIndentRegex or decreaseIndentRegex
+
+    if decreaseNextIndentRegex?.testSync(precedingLine) or (decreaseIndentRegex?.testSync(tokenizedLine.text) and not @editor.isBufferRowCommented(bufferRow))
+      lastIndentingRow = bufferRow
+      while (lastIndentingRow = @buffer.previousNonBlankRow(lastIndentingRow))?
+        continue if @editor.isBufferRowCommented(lastIndentingRow)
 
-    return desiredIndentLevel unless decreaseIndentRegex = @decreaseIndentRegexForScopeDescriptor(scopeDescriptor)
-    desiredIndentLevel -= 1 if decreaseIndentRegex.testSync(tokenizedLine.text)
+        lastIndentingLine = @buffer.lineForRow(lastIndentingRow)
+        if increaseNextIndentRegex.testSync(lastIndentingLine)
+          desiredIndentLevel = @editor.indentationForBufferRow(lastIndentingRow) + 1
+
+          for row in [lastIndentingRow..bufferRow]
+            continue if @editor.isBufferRowCommented(row)
+
+            rowLine = @buffer.lineForRow(row)
+            desiredIndentLevel -= 1 if decreaseNextIndentRegex?.testSync(rowLine)
+            desiredIndentLevel -= 1 if decreaseIndentRegex?.testSync(rowLine)
+
+          return Math.max(desiredIndentLevel, 0)
```

Comment:
I don't fully understand what this new code does, and I'm a little bit concerned about the performance implications of calling `testSync()` in doubly-nested loop. Just to help me understand: could you explain in a little more detail your intent behind the `while` loop and the inner `for` loop?